### PR TITLE
Extract subprocess API used in rz-test into RzUtil

### PR DESCRIPTION
**Pull request by ret2libc**
Do not squash.

### CODE REVIEW

File: librz/util/sys.c [hunk 1]
```diff
@@ -1380,3 +1383,248 @@ RZ_API void rz_sys_info_free(RSysInfo *si) {
...
+
+/**
+ * \brief Close a file descriptor previously created pipe \p rz_sys_pipe
+ */
+RZ_API int rz_sys_pipe_close(int fd) {
+	return close (fd);
+}
+
+RZ_API int rz_sys_execv(const char *pathname, char *const argv[]) {
+	rz_th_lock_enter (sys_pipe_mutex);
```
Comment (thestr4ng3r) [inline librz/util/sys.c]: Is this really necessary around the `execv` calls? I don't really see any races here.

File: librz/util/sys.c [hunk 2]
```diff
@@ -1380,3 +1383,248 @@ RZ_API void rz_sys_info_free(RSysInfo *si) {
...
+	bool deleted = ht_uu_delete (fd2close, fd);
+	rz_warn_if_fail (deleted);
+	int res = close (fd);
+	rz_th_lock_leave (sys_pipe_mutex);
+	return res;
+}
+
+RZ_API int rz_sys_execv(const char *pathname, char *const argv[]) {
+	rz_th_lock_enter (sys_pipe_mutex);
```
Comment (thestr4ng3r) [inline librz/util/sys.c:1623]: Here it makes sense of course since you call `close_fds()`.

File: librz/util/sys.c (hunk 1 above)
Comment (ret2libc) [reply librz/util/sys.c]: I think this is necessary to ensure an `execv` is not executed in between the `open` and the `fcntl`. If that would happen, you would have a pipe fd that leaks in the child process.

Comment (thestr4ng3r) [reply librz/util/sys.c]: Good point, maybe add a comment about this since it's a bit non-intuitive that locking around a single syscall is necessary.

**Review by XVilka — COMMENTED** [review]
Overall is good, but we need to make sure it works in all supported platforms .

**Comment by XVilka** [comment]
ACR (configure/make) build is broken:
```
gcc -c    -MD   -fPIC -g -Wall -D__UNIX__=1  -MD   -fPIC -g -Wall -D__UNIX__=1 -DRZ_PLUGIN_INCORE -I/home/runner/work/rizin/rizin/librz/..//shlr -I/home/runner/work/rizin/rizin/librz/../shlr/sdb/src -I/home/runner/work/rizin/rizin/librz/../shlr/zip/include -I/home/runner/work/rizin/rizin/librz -I/home/runner/work/rizin/rizin/librz/include -I/home/runner/work/rizin/rizin/librz/../shlr/sdb/src -fvisibility=hidden -I../../shlr/sdb/src/ -DHAVE_FORK=1 -DUSE_R2=1 -I../../shlr  -MD   -fPIC -g -Wall -D__UNIX__=1 -DRZ_PLUGIN_INCORE -I/home/runner/work/rizin/rizin/librz/..//shlr -I/home/runner/work/rizin/rizin/librz/../shlr/sdb/src -I/home/runner/work/rizin/rizin/librz/../shlr/zip/include -I/home/runner/work/rizin/rizin/librz -I/home/runner/work/rizin/rizin/librz/include -I/home/runner/work/rizin/rizin/librz/../shlr/sdb/src -fvisibility=hidden -I../../shlr/sdb/src/ -DHAVE_FORK=1 -DUSE_R2=1 -I../../shlr -o ../../shlr/spp/spp.o ../../shlr/spp/spp.c
In file included from /home/runner/work/rizin/rizin/librz/include/rz_types.h:8:0,
/home/runner/work/rizin/rizin/librz/../global.mk:38: recipe for target '../../shlr/spp/spp.o' failed
                 from /home/runner/work/rizin/rizin/librz/include/rz_util.h:6,
                 from ../../shlr/spp/spp.h:81,
                 from ../../shlr/spp/spp.c:3:
/home/runner/work/rizin/rizin/librz/include/rz_userconf.h:18:20: error: token "@" is not valid in preprocessor expressions
 #define HAVE_EXECL @HAVE_EXECL@
                    ^
/home/runner/work/rizin/rizin/librz/include/rz_util/rz_sys.h:55:17: note: in expansion of macro ‘HAVE_EXECL’
 #if __UNIX__ && HAVE_EXECL && HAVE_PIPE2
                 ^~~~~~~~~~
In file included from ../../shlr/spp/spp.c:5:0:
../../shlr/spp/config.h:9:0: warning: "HAVE_SYSTEM" redefined
 #  define HAVE_SYSTEM 1
 
In file included from /home/runner/work/rizin/rizin/librz/include/rz_types.h:8:0,
                 from /home/runner/work/rizin/rizin/librz/include/rz_util.h:6,
                 from ../../shlr/spp/spp.h:81,
                 from ../../shlr/spp/spp.c:3:
/home/runner/work/rizin/rizin/librz/include/rz_userconf.h:19:0: note: this is the location of the previous definition
 #define HAVE_SYSTEM @HAVE_SYSTEM@
```

**Comment by ret2libc** [comment]
I know I'm already working on the fix, it's very simple. And yes, it should work on all platforms as i have provided different implementations based on availability of functions and o_cloexec.

**Review by thestr4ng3r — DISMISSED** [review]

**Review by thestr4ng3r — COMMENTED** [review]
Wait, why revert these O_CLOEXEC things?

**Comment by XVilka** [comment]
You seem downgraded the ACR version to generate `.configure` script. Could you please update your acr and restore the "quiet" mode? While ACR/Makefiles are not removed completely we should still maintain them.

File: librz/util/subprocess.c [hunk 3]
```diff
@@ -0,0 +1,791 @@
...
+}
+
+RZ_API char *rz_subprocess_out(RzSubprocess *proc) {
+	return rz_strbuf_drain_nofree (&proc->out);
+}
+
+RZ_API char *rz_subprocess_err(RzSubprocess *proc) {
+	return rz_strbuf_drain_nofree (&proc->err);
+}
+#else
```
Comment (XVilka) [inline librz/util/subprocess.c]: Please add here a comment of what `#if[def]` branch it's end since the blocks are big.

**Review by XVilka — CHANGES_REQUESTED** [review]
So far looks good now 👍

**Review by XVilka — APPROVED** [review]
Looks good now and works everywhere it seems.

On the side note - it seems to made Windows testing much faster 👍 🚀

**Review by thestr4ng3r — APPROVED** [review]

**Pull request merged**